### DISCUSSION TITLE
How can I use this in nixos?

### DISCUSSION THREAD

**Discussion by liozur** [prasanthrangan/hyprdots#166] [Q&A]
I just recently discovered hyprland, and this excellent repository. Since I'm a novice in this area, a quick question: is it possible to use this repository on nixos?

**Comment by prasanthrangan** [comment]
Hi, thank you!

It should be possible in theory, but I'm not sure how nix configs work as I'm kinda new to Linux as well and only (mostly) used Arch.

If anyone knows how to re-configure this in nix, fill free to copy my dots to create a nix version!

**Reply by bphd** [reply]
(and to share the link into that discussion)

**Reply by Scion0Roses** [reply]
nixos is amazing, similar to flatpak in dependency management, additionally a similar project to flathub: https://mynixos.com/. 

I would suggest a simple HyDE.nix that installs/wraps all deps, then simply copy over the dotfiles same as arch. System configs like mkinitcpio are also done in nix files. 

Warning about home manager, it restricts access to the home folder. <suck> (Most people redirect their dotfiles anyway, I don't... Vanilla supreme)

Avoid experimental features like flakes, version control won't be necessary for nixos users.

Consult:
https://search.nixos.org/options
https://search.nixos.org/packages?

**Reply by sandeng1440** [reply]
Hi @kyn-0s1r15, I was able to do it with a flake. Basically, it's what you said: install Hyde-cli, then use Hyde-cli to restore the dotfiles. [Here is my repo for the flake.](https://github.com/sandeng1440/nixos-config). The Readme will guide you on how to set it up.

**Comment by Ashe** [comment]
I'm working on this - https://github.com/ashe/dotfiles/tree/theme-switching

Note that due to the nature of my config its heavily specialised to my system. That said, if you look at `home/modules/themes` you can see that it's an optional module that hooks into my other modules, ammending config files etc.

Some things will need scripts to get around; others need some home-manager hackery. It's a slow process. Right now, GTK is the most annoying thing (maybe this is why hyprdots prefers dolphin over nautilus, food for thought :stuck_out_tongue: ). I think it's because nautilus is on GTK 4 maybe? And I've only configured GTK 3? I have no clue. 

Will probably just move on to QT soon and then learn about Kvantum and how that works. Lots to do.

**Comment by prasanthrangan** [comment]
> Some things will need scripts to get around; others need some home-manager hackery. It's a slow process. Right now, GTK is the most annoying thing (maybe this is why hyprdots prefers dolphin over nautilus, food for thought 😛 ). I think it's because nautilus is on GTK 4 maybe? And I've only configured GTK 3? I have no clue.

hey sorry I have not used NIX, but all the scripts I use here are very much kinda basic, so these scripts should work in NIX as well right?!?

> Will probably just move on to QT soon and then learn about Kvantum and how that works. Lots to do.

Kvantum is completely svg file based engine, so its highly customizable, easy to manually modify from an existing theme but difficult to generate it on the fly!

**Reply by Ashe** [reply]
Largely yes! You've done very well :) I don't really do a lot of bash so I'm thankful you've done some heavy lifting! Most of the stuff I'm adding is to simply adapt it. For instance, in my nix config I have an absolute list of themes and their associated data, and I wanted to adapt your scripts to use this centralized data source, meaning that I can control which themes are enabled from my nix module as well as there being some barriers in place if any of the required config files are missing (you've got lots of colourschemes and lots of programs to theme, I'm definitely going to miss a few!)

Also, configuration in nix is largely immutable. I can intercept it a few ways when my theming module is invoked:

1. If the configuration of the program allows for including other files, I just squeak in an include to a file operated on by your scripts. Then, instead of doing your replace-in-file code, I'm generating a partial config with just the theme-specific data.

2. If the configuration doesn't allow for including files, I redirect the my config to a different file. Then, I've edited your script to take my base config and append the theme-specific stuff to it.

So its mostly the output of your scripts I'm changing, not so much the logic. Taking it one step at a time.

Since Ive seen a few posts about nix and hyprdots, perhaps when I've finished / got as far as I'm comfortable we can try to streamline the process for other users, but for now I'm just focusing on finding a way that works with my system as it'll get a lot more complex trying to make it generic (i've not done anything like that before).

Thank you for your great work :)

**Reply by prasanthrangan** [reply]
> Largely yes! You've done very well :) I don't really do a lot of bash so I'm thankful you've done some heavy lifting! Most of the stuff I'm adding is to simply adapt it. For instance, in my nix config I have an absolute list of themes and their associated data, and I wanted to adapt your scripts to use this centralized data source, meaning that I can control which themes are enabled from my nix module as well as there being some barriers in place if any of the required config files are missing (you've got lots of colourschemes and lots of programs to theme, I'm definitely going to miss a few!)

Thank you so much, for your support! please refer [theming.md](https://github.com/prasanthrangan/hyprdots/blob/main/theming.md) for the config file structure. (its a bit outdated, but I'll update it when i get some time)


> Also, configuration in nix is largely immutable. I can intercept it a few ways when my theming module is invoked:
> 
>  1. If the configuration of the program allows for including other files, I just squeak in an include to a file operated on by your scripts. Then, instead of doing your replace-in-file code, I'm generating a partial config with just the theme-specific data.
> 
>   2. If the configuration doesn't allow for including files, I redirect the my config to a different file. Then, I've edited your script to take my base config and append the theme-specific stuff to it.
> 
> 
> So its mostly the output of your scripts I'm changing, not so much the logic. Taking it one step at a time.
> 
> Since Ive seen a few posts about nix and hyprdots, perhaps when I've finished / got as far as I'm comfortable we can try to streamline the process for other users, but for now I'm just focusing on finding a way that works with my system as it'll get a lot more complex trying to make it generic (i've not done anything like that before).

sounds great... once you configure one theme to work on your system, the others should be simple to implement, so no worries... Also let us know once you are able to get this working and I'll link your repo, so user who prefer NIX can use your version!

**Reply by Ashe** [reply]
Current progress with rofi looks like this: 

![image](https://github.com/prasanthrangan/hyprdots/assets/8940077/d05a480b-9697-4d46-b993-72c464a32c1f)

I have no idea what's going on honestly; I'm using your rofi config and even though the wallpaper is appearing the list box is not.. If I delete "inputbar" from the children of mainbox, or if set enabled to false for inputbar, then things start to work. Unsure if it's a bug or what. Little bit weird as it doesn't seem logical. No errors when running from the command line either.

Style 2 functions a lot better thankfully!
![image](https://github.com/prasanthrangan/hyprdots/assets/8940077/01039f66-571c-44fb-823d-0fad1f172846)

I'll stick with style 2 and crack on with waybar next. I've been waiting a long time to redo my horrific looking bar so looking forward to seeing what we can do!

I am looking to implement rofi theme switching and making use of the import functionality. I'll essentially have all styles loaded in my nix config and then use a script to glue it together. I need at least 1 working style for that to happen though and it's low on the priority list so for now I'm just assuming 1 rofi config :)

---

> Also let us know once you are able to get this working and I'll link your repo

Don't get your hopes up, I don't think I'll be able to 100% implement everythign your dotfiles implements. I love your theme switching for terminals and your rofi/waybar (and probably dunst) coordinated themes and want to take inspiration (read: *steal* :stuck_out_tongue: ) them. I did actually implement the wallbash stuff as I didn't really understand the point; when I actually got to the part where my colourscheme can get set via the wallpaper I actually realised that's not quite what I'm looking for. Don't get me wrong, that's a cool feature, but my reasoning is that that's a lot of work for me to maintain.

The nix way of doing things would be to do a lot of this work at compile time. I'm using your scripts where it's just doing some simple editing, but when the majority of the theming comes from stuff like wallbash then that's when it's less in the hands of my nix config. You *could* fix this by listing your wallpapers in nix, but I like the idea that I can just plop in whatever wallpapers I like.


The magic is all in [this file](https://github.com/Ashe/dotfiles/blob/theme-switching/home/modules/themes/default.nix). You can see here that I've 
1. Listed my themes
2. Sym linked scripts and wallpapers etc
3. Done config-dependent customisation to inject themeing into my other config.

None of this is drop-in so I'd hesitate to say someone can just copy what I've done. That said hopefully it's a good starting point for someone :)

---

So, tl;dr, I'm not implementing *everything* and I'm sure someone will do a 100% faithful recreation of your dotfiles (and maybe they will find mine a good starting point?). I like the idea that I have lots of themes and that nix will check if I've missed any files when I try to regenerate my config, but the more of your stuff I implement the more I'm relying on bash and not nix. 

Thanks as always for your hard work :) if you know why that rofi thing might be an issue please let me know!

**Reply by bphd** [reply]
Which commit and lines concern HyprDots?

**Reply by Ashe** [reply]
@bphd everything in that linked file on that branch is related to hyprdots. That's the starting point. Then you have the adjacent script directory, as well as all the bespoke methods for injecting themes into my config.

Again, this is *not* a general thing, it's just how I've done it for my config. I don't have the know-how to do this in a way which satisfies everyone's configs especially when the home-manager modules are intentionally immutable. You need to essentially add an include that is *not* in your config, that you then generate separately using `switch-theme.sh` (my version of whatever it was called in hyprdots).

**Reply by bphd** [reply]
@Ashe, sure it's okay for it not being a general thing, even if it's Ashe's HyperDots it's okay. I've just wanted to make sure to not apply things that really have nothing to do with the work you've did related to Prasan's HyperDots

**Comment by prasanthrangan** [comment]
> I have no idea what's going on honestly; I'm using your rofi config and even though the wallpaper is appearing the list box is not.. If I delete "inputbar" from the children of mainbox, or if set enabled to false for inputbar, then things start to work. Unsure if it's a bug or what. Little bit weird as it doesn't seem logical. No errors when running from the command line either.
I'll stick with style 2 and crack on with waybar next. I've been waiting a long time to redo my horrific looking bar so looking forward to seeing what we can do!
I am looking to implement rofi theme switching and making use of the import functionality. I'll essentially have all styles loaded in my nix config and then use a script to glue it together. I need at least 1 working style for that to happen though and it's low on the priority list so for now I'm just assuming 1 rofi config :)

oh strange, does it only happen with style_1? my wild guess would be the `width` set for `inputbar` for style_1 -
https://github.com/prasanthrangan/hyprdots/blob/d2e8cfb0f913f8c8dc6de24b5cf8141b260bef2c/Configs/.config/rofi/styles/style_1.rasi#L46
so, could you please try adjusting (reducing) this value (also try using `px` instead of `%`) to scale it for your monitor res...


> Don't get your hopes up, I don't think I'll be able to 100% implement everythign your dotfiles implements. I love your theme switching for terminals and your rofi/waybar (and probably dunst) coordinated themes and want to take inspiration (read: _steal_ 😛 ) them. I did actually implement the wallbash stuff as I didn't really understand the point; when I actually got to the part where my colourscheme can get set via the wallpaper I actually realised that's not quite what I'm looking for. Don't get me wrong, that's a cool feature, but my reasoning is that that's a lot of work for me to maintain.

Hahaa no worries... I didn't mean 100% feature parity, i was Just referring to wallpaper and theme switching.
Also wallbash is very rough and hacky way of doing things, so not an elegant solution... but hey, it works for me!

> The nix way of doing things would be to do a lot of this work at compile time. I'm using your scripts where it's just doing some simple editing, but when the majority of the theming comes from stuff like wallbash then that's when it's less in the hands of my nix config. You _could_ fix this by listing your wallpapers in nix, but I like the idea that I can just plop in whatever wallpapers I like.
> 
> The magic is all in [this file](https://github.com/Ashe/dotfiles/blob/theme-switching/home/modules/themes/default.nix). You can see here that I've
>  1. Listed my themes
>  2. Sym linked scripts and wallpapers etc
>  3. Done config-dependent customisation to inject themeing into my other config.
> 
> None of this is drop-in so I'd hesitate to say someone can just copy what I've done. That said hopefully it's a good starting point for someone :)

Oof... Looks like I underestimated the work required to configure nix... But your config files looks good, so looking forward to see it working in your system :)

**Reply by Cu3PO42** [reply]
> Oof... Looks like I underestimated the work required to configure nix... But your config files looks good, so looking forward to see it working in your system :)

As I have learned myself, it's quite the undertaking. Initially, I started with your repo and started porting things, but have now switched over to starting from scratch and picking bit after bit and getting it to work. The other approach might work to, but this way I'm actually learning all about GTK and Qt theming, so that's a bonus for me.

I'm sure there is a much easier and direct route to "just getting Hyprdots to work" on NixOS, it's just that I want to do things the NixOS/Home-Manager way, which is partly incompatible with... pretty much every other way.

I have arrived at roughly the same setup that @Ashe also has - or at least that's what I'm planning, I'm not quite there yet.
I'm trying to design my config in a way that it should be relatively easy to re-use by others and hope to write a guide for that in the future, which I'll definitely link here.

@Ashe if you'd like to chat about the porting and dynamic theme switching, I can be reached on Matrix (Cu3PO42@matrix.org) or Discord (@Cu3PO42).

**Reply by Ashe** [reply]
Thank you for the offer but my dotfiles are in a place I'm happy with; I have a way of switching my colourscheme, wallpaper and bar on the fly. It's not a full port but I'm also just busy with other things. My main objective was to be able to change my colourscheme without rebuilding my config and hyprdots already had collected a nice set of cohesive colourschemes and wallpapers.

Thank you to everyone involved :)

**Reply by Cu3PO42** [reply]
Being happy with my dotfiles is definitely something I aspire to be some day, so I salute you! That being the case, I might yoink a few bits from your config should I get stuck.

**Comment by lavara123** [comment]
What's the latest situation? Is the same for NixOS?

**Reply by Cu3PO42** [reply]
To my knowledge nothing has changed. Ashe has ported parts of the functionality to her dotfiles (which are linked above) and I am currently working on my own port. However, my work is still very much in progress and not yet public. I do, however, aim to produce NixOS/Home-Manager modules that you can just drop into your config. I'll post here once I have gotten to a state I like.

**Comment by Cu3PO42** [comment]
Hey all! I want to give a quick overview of my progarss towards replicating a similar feature set on my NixOS system with Home-Manager.

First off, let me say that the amount of effort that must have gone into Hyprdots is quite astonishing and I do appreciate your work.

I initiall thought I could get away with checking out the repo, applying a few patches and rewriting the installation scripts.
That effort has proven infeasible or at the very least impractical.

The problem is that there's only one thing that Home-Manager dislikes more than state and that is state mixed with configuration.
Unfortunately, the Hyprdots' scripts heavily rely on mutating files, including the `.ctl` files that both indicate which of a number of configurations is active and the configurations themselves. To make proper use of Home-Manager, you need to keep as many things as possible read-only and carefuly isolate state. As such I have chosen to reimplement a lot of functionality, rather than port it.

## Core Approach

### Chroma

I rely on an approach similar to, but still somewhat different from, Ashe's. The core of my efforts is my new theming module that I've dubbed *Chroma*. You can find it in [copper-chroma.nix](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/modules/home-manager/copper-chroma.nix). It exposes a theming system that other apps and frameworks can be plugged into. Integrations for supported applications are then defined in the [copper-chroma folder](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/modules/home-manager/copper-chroma). Theme definitions are not distributed among the various application configs, but centralized in the [themes](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/themes) directory.

From this definition, I build one folder per theme, containing all relevant files and settings. For configurations that support includes, I only include the theme config, for tools whose configs do not support includes, I template the whole configuration. All tools are configured to read configs from `~/.config/chroma/active/<tool>`, where `active` is a symlink to one of the theme folders. Theme switching then consists of updating the symlink and running an automatically generated activation script. My equivalent of `themeswitch.sh` is [`chromactl`](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/packages/chromactl/chromactl.sh).

### Swim

I have also built wallpaper switching from the ground-up as my [*swim* module](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/modules/home-manager/copper-swim.nix). Swim can hook into Chroma to offer a different set of wallpapers per theme. Due to the modularity of Chroma, it needs no special knowledge of Swim and the only Chroma-specific code for Swim is remembering the last-active wallpaper for each theme. Neither of the above know anything about Rofi or the thumbnails it needs.

## Further Differences

Wherever possible, I strive to isolate features from one another as much as possible. You can pick and choose which parts of Hyprdots you want to activate, by activating my corresponding [features](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/features/home-manager/nixos/).

I have tried to keep any large assets out of my main repository, that means I install all themes from a package: preferably from the standard sources, though I have also packaged themes that weren't already available.

Luckily, most themes are already packaged in nixpkgs, so I didn't need to do significant work for them. One that was not already available is the SDDM theme, which I have packaged in [`sddm-theme-corners.nix`](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/packages/sddm-theme-corners.nix). Essentially, it fetches the theme from the upstream GitHub repo and specifies the necessary dependencies.

My Hyprland initialization is quite different from upstream. Instead of launching dunst, waybar, etc. as part of my Hyprland configuration, I manage all of these via systemd user units. Also, the `scripts` directory is gone. Some scripts are replaced entirly by my own solutions, as outlined above and some others I have structured into various [`packages`](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/packages/). Scripts coming from Hyprdots are in folders ending in `-hyprdots`.

## Feature support

As of now, this is an incomplete port/rewrite, and it will never be complete in the sense that there's some features I don't plan to port.
However, there are some things I still want to port:
* wallbash, i.e. dynamic theming based on the current wallpaper,
* theming of VSCode
* media player support in Waybar,
* GPU usage support in Waybar.

Among the things I don't plan to support for the forseeable future is:
* GRUB theming, as my system relies on systemd-boot
* Spotify theming, as I use Apple Music
* Steam theming, as I still dual-boot Windows for gaming
* Hyprdot's ZSH configuration, as I use my own Fish config
* Automatically sycing SDDM wallpaper to current desktop wallpaper
* Possibly more that I'm forgetting

Thare are also some things, which I plan that are not yet in Hyprdots:
* Support theming of Flatpak Qt apps,
* Solid GTK 4 theming,
* More themes!

## Advantages

There are two main advantages to my approach: it is easily extensible, i.e. someone could easily maintain theming for their own set of applications without needing to maintain a fork and someone else could easily have their own local themes.

It is also easier to cross-reference theme settings without complicated scripts. For example, I can easily set the Qt font based on the same value as the Gtk font without any need to parse configuration files, since everything is available in Nix.

### Usage under non-NixOS
Further, the majority of my configuration work is specific to Home-Manager and can be used on non-NixOS Linux distributions that uses systemd. This should get you a majority of the way, even if Hyprdots doesn't dirctly support your distribution of choice. You'll loose out on anything configured in the [NixOS module](https://github.com/Cu3PO42/gleaming-glacier/blob/feat/chroma/features/nixos/hyprland.nix).
In particular that means, you need to:
* install hyprland, xdg-desktop-portal-hypr, xdg-desktop-portal-gtk, swaylock-effects, and swayidle,
* configure SDDM to your preferred theme yourself,
* setup blueman,
* make sure dconf is available on your system.

## Future Outlook

I'm not nearly done yet. I'd like to clean up my configuration some more, verify that everything works as intended, and write documenation. When I'm done with all that, I will probably make some more adjustments so that this is not just Hyprdots but for Nix, but rather a mix of Hyprdots and my own preferences.

Of those tools and configurations that I ultimately keep, I'd like to rewrite their config generation scripts. While I have kept them mostly as is for now, a lot of the work that happens in `wlogoutlaunch.sh`, for example, could be handled purely in Nix at build-time. In my opinion, that would greatly simplify things and make them less error-prone.

## Questions

Finally, I'd also like to use this opportunity to ask some questions, which I haven't yet achieved clarity on. I'm hoping you can provide your insight, @prasanthrangan:

* Is it correct that the Rofi instances launched by `themeselect.sh`, `swwwwallselect.sh`, etc. take up the entire width of the screen?
* I currently use your configuration of swaylock exactly. However, rather than going into sleep after some time, my PC remains on and only goes to sleep once I unlock swaylock. Have you observed that before?
* Dunst is, as far as I can tell, only themed with the icon theme. Does color theming happen automatically or is this simply not yet implemented?

**Reply by Cu3PO42** [reply]
Hello everyone! I'd like to give you another update on my efforts. Not too much has changed over the last weeks, as I actually took some vacation time. And though I haven't pushed anything, I have, for example, gotten libadwaita theming to work consistently and made some configuration more flexible (e.g. Qt now always has the same font as GTK). None of that is published yet, but it should be very soon.

Unfortunately, I have come to realize that while Hyprdots is great and a tremendous effort, for a variety of reasons, I will not pursue a straight up port of Hyprdots to NixOS/Home-Manager anymore.

* My configuration has diverged so far from Hyprdots that keeping them in sync requires more effort than I can justify.
* As great as this configuration is, I really want 'my own', which does not only mean some configuration overrides, but entirely different tools in some places.

I want to truly thank everyone involved with the project for their work and wish you the best with the continued effort. It has gotten me off the ground and inspired many design decisions I have made, and some parts of Hyprdots will continue on in my own configuration.

For anyone else on NixOS/Home-Manager, you are welcome to look at [my configuration](https://github.com/Cu3PO42/gleaming-glacier/tree/feat/chroma) and keep an eye out for future updates. I aim to make it easy to include in other Flakes, but I do not yet have documentation for all of the new theming modules. And again, it's not a complete port of Hyprdots, but rather my own configuration that uses some parts of Hyprdots.

**Reply by Cu3PO42** [reply]
All of the updates I talked about are now available on [my repo](https://github.com/Cu3PO42/gleaming-glacier/tree/feat/chroma)! I've also made some minor adjustments/fixes to my port of Hyprdots functionality. However, this is likely the last of Hyprdots-related updates for now :-(

**Reply by maxbol** [reply]
@Cu3PO42 Did you remove the chroma repo? I don't seem to able to access it.

I'm in the exact same situation as you and seeing the same issues with nixifying, so even getting to look at your conf as a reference would be worth a lot. I might even be able to do a PR on some stuff that you are not prioritizing - like Spotify themeing :)

**Reply by Cu3PO42** [reply]
@maxbol Not as far as I can tell. The links above still work for me, even in a private window. However, the linked branch is now out-of-date. I'd recommend you look at [`next`](https://github.com/Cu3PO42/gleaming-glacier/tree/next) instead, where I've made other major updates to the Flake infrastructure. I'm currently in the process of writing docs for all of that, but they, unfortunately, aren't ready yet. I expect that as soon as they are, `next` will become the new `master`.

I'm generally happy to accept contributions, though Spotify in particular is not something I'm using, so I'm not sure about having that upstream. That said, Chroma is designed to be fully extensible from your end without a fork, so that shouldn't pose any problem. Feel free to shoot me a message about features you're missing and we can discuss how to best add them.

**Reply by maxbol** [reply]
Sorry, seems like there was some issue with the GH mobile app. Working as expected on desktop :)

**Comment by bphd** [comment]
Headed here to see progress as I hadn't the time to port it myself and @Cu3PO42 has made something really good to think the project "statelessly" or idk how to name that, but it has become such hydra that I truly don't know how to handle it, worse, now it's semi-abandoned as it strongly derived and integrated own dotfiles to become something totally different. Though the theming project it had gave birth to is really promising, it has to find it's public now. @Ashe has (I think) not much changed since and is like a subset integrated into dotfiles without a clear way of what and how to take it to implement into self systems. So so far this theme still is a dream for me ahah

**Reply by Cu3PO42** [reply]
My dotfiles aren't abandoned at all. But I have merged Chroma into a new branch, thus new stuff is happening there rather than the branch linked above. It's called [`next`](https://github.com/Cu3PO42/gleaming-glacier/tree/next). You're correct, however, that they're my dofiles, not a 1:1 port of Hyprdots. You can grab just the parts you like, but the documentation is currently very out-of-date. I'll write new documentation at some point, but no promises when that is going to be.

**Reply by bphd** [reply]
I talk about "semi-abandoned" talking about the more "authentic" port made before `next`, while/before `chroma`. But you've already stated that you've stopped working on HyprDots as is. There is bits that still are part of your configuration but it's not actively maintained, just some remainings like. And about grabbing part that I like, the project is such an hydra in different functionalities and own logic that I struggle hard to know what to do to import what your project has as HyprDots*ish* to my configuration.nix. Maybe you have some insights?

**Comment by kik4444** [comment]
Has anyone tried simply modifying the file paths in all the scripts and configs that reference files that don't exist in NixOS (e.g. anything in `/usr/*`)? I imagine doing that would be easier than trying to port everything to the proper Nix way with Home-Manager and so on.

**Reply by bphd** [reply]
@kik4444 Haven't tried but maybe the problem is wider than that, like reproducible OS don't tend to execute scripts per default, or really anything that isn't installed by the package manager. But if you could provide an example patch/MR/commit to show what exactly could be done that would be wonderful

**Reply by kik4444** [reply]
I suppose the closest thing would be https://github.com/prasanthrangan/hyprdots/pull/1141, but I haven't done anything important so far.

I know that in NixOS you're not supposed to run random scripts to modify system files, but your home dir is free real estate imo (also `/tmp`). I think that as long as hyprdots is patched to use file paths that exist in both NixOS and Arch within your home dir, it *could* work the same as it does on Arch without making many significant changes.

**Reply by bphd** [reply]
So you think that helping the os find `bash` would be enough for everything to be executed alright? That script don't touch anything outside `/home` folder? Not even a general configuration file telling the system to go look in `/home` or something? If it's really only about that so yeah, transactionify the project could be overkill (but at least smooth, if someone doesn't bother doing it). Thing is that till here nobody dared doing it for HyprDots, people only takes inspiration here and there but what they is not something that you import and it'll only apply HyprDots, at least they didn't bothered isolating it in their process. It's rather something that modifies a lot of the system and that didn't even takes much from HyprDots, and less and less over time. So yeah that could make it at least work in that way

Random example but [here](https://github.com/prasanthrangan/hyprdots/blob/26f8e960eb630c6dde6cbd248022ec7c79647116/Scripts/.old/zsh_plugin_installer.py#L22) it doesn't seem to be an especially universal path https://github.com/prasanthrangan/hyprdots/blob/26f8e960eb630c6dde6cbd248022ec7c79647116/Scripts/.old/zsh_plugin_installer.py#L22

**Reply by kik4444** [reply]
I'm not gonna bother with cases like that because I don't use zsh anymore and when I did use it I preferred writing my own zshrc without oh my zsh. I'm strictly focusing on the Configs folder

**Reply by bphd** [reply]
That's the reason why tweaks of people aren't drag and droppable, far from that. They only takes part that they like, worse, they tweak it at their taste. At least, a nice thing would be to do the work in two part: 1. Applying changes to universify parts that we are willing to port (at least for our needs). Referencing somewhere that change. 2. Making a second change to our taste. So people could take only the first one. Better, if that first step could be merged to upstream so the project ifself is a bit more compatible over time

**Reply by kik4444** [reply]
Speaking of tweaks do you happen to know why `Scripts/restore_fnt.lst` extracts archives to both `/usr/share/...` and `$HOME/...`? Do the home paths alone not suffice? The /usr paths are off limits in NixOS so I'm not sure how to solve this one without potentially breaking something important. Is it safe to delete the `/usr/share/...` lines? Or should I add logic to `Scripts/restore_fnt.sh` to skip that dir on NixOS?

**Reply by kik4444** [reply]
Honestly, having come this far, I have some understanding of how Hyprdots works overall. At this point it would be easier to just take what I want and go my own way. But I've already come this far and now I'm just curious to see what will happen if I try running hyprdots as is on NixOS with my changes. I just wanna see if this thing blows up or ends up working 🙂

**Reply by kRHYME7** [reply]
@kik4444 If you managed to do it on Nixos I'll be trying nix on one of my machine. I guess it would be cool if we have a  port Hyprdots for NixOS, if possible.


Thanks in advance!
I have a silly question, does home-manager have the same syntax and configuration as NixOS? I barely scratch the surface of nix packages and home-manager,and I would still prefer the mainstream Linux infrastructure on Arch. So I think I can only commit to using home-manger as of now.

**Reply by kik4444** [reply]
It's still the nix language at the end of the day so it's the same. Though the way you structure the config between a full system and home-manager standalone will likely be different.

**Reply by kik4444** [reply]
Here's my experience with the last patches I made https://github.com/prasanthrangan/hyprdots/pull/1152

**Reply by bphd** [reply]
Nix Is thought at core to be break immune, little bit like Fedora silverblue. Would be utterly funny that it doesn't survive such operations. As long as you don't touch the read only mounts, the system should be able to reproduce itself to identical state. At least it's how it's sold and why much people use it

Though idk the logic behind implementing inside and outside of home, maybe @prasanthrangan should know. But afaik, home conf have priority over any other and /usr ones tend to be the default one, multi user one, root ones. It's generally way more reliable to apply those in multiple place in case something doesn't care about home configuration. More than that idk, have to check on a software basis manual to know when and where and if they take into account home configurations. Afaik you can even generate a VM machine upon recompiling of the OS to test on a safe environment that still is pretty much a 1:1 to yours to be sure about compatibility (considering you don't tweak outside of NixOS configurations files)

@kRHYME7 You can install HM and NixOS aswell on any other OSes, it's a package manager first and foremost. Like actually I use Nix package manager on Guix when I'm too bored of porting something, because even if Guix follows the Nix logic it's like really more complicated to port things. For example for HyprDots, I'd have to find how to convert SystemD hooks to SheperD ones. Though SheperD is really more close to POSIX logic than SystemD

**Comment by kRHYME7** [comment]
Hello let's assume we can fork and maintain this project with  Home Manager( Will try to use it again in arch). Is this compatible with (lesser adjustments) for the whole nix OS?

Also having it in HM, by theory we can install this easily in any distro including debian base,right?.

**Comment by ghostarray263** [comment]
Can you use this in arco Linux?

**Reply by kRHYME7** [reply]
It should work on any distro that uses ` pacman ` command

**Reply by ghostarray263** [reply]
Can't I download all files that use git I have bad network so when I tried installing I had errors and sometimes not because of network I didn't know what was even wrong can you help me figure it out? I can give you my telegram or something please sir

**Reply by kRHYME7** [reply]
We got a Discord channel.  Check it out in README. 

Here's how we can install it 
* Download the zip file  
![image](https://github.com/user-attachments/assets/34c54f56-81f4-4ca6-b8d8-6c92343cc422)
* Extract it somwhere (preferrably at ~/HyDE )
* you can use the CLI by ` Hyde-install --dir ~/HyDE ` (~/HyDE is the directory of the Clone extracted repo ) 
* Follow the instructions
* ONLY choose a SINGLE theme in case git becomes slow

**Reply by ghostarray263** [reply]
Sorry to ask so many questions but what do you mean by SINGLE and also git for me don't work mostly

**Reply by kRHYME7** [reply]
My bad I meant Single theme
Themes are installed from different repo, so I just highlighted that only choose a single theme. 

While in installation it will ask you this.
![image](https://github.com/user-attachments/assets/b2687d7d-7492-406c-ac93-e007c4ffe3de)
Choose only a single theme for now so you got lesser problems with the rest of the installation

**Comment by sandeng1440** [comment]
Hello, I have just succeeded in making hyprdots work in nixos. Here's my [flake](https://github.com/sandeng1440/nixos-config) if you are interested.

**Comment by bphd** [comment]
Guix here

**Comment by richen604** [comment]
I'm working on the home-manager nixos implementation [here @ hyprdots-nix](https://github.com/richen604/hyprdots-nix). currently supports being run as a vm for easier development

decided to support both the standard @kRHYME7 's hyde-cli installation and home-manager as there's some caveats around the proper nix way to do things. 

hyde-cli method will be setup soon. for now @sandeng1440 has a good way.
